### PR TITLE
feat: Collection offset implementation

### LINKED ISSUE #1319: [feature request] Add time offset option to input plugins
### Proposal:

I would like to define the precise time when an input plugin is being executed. Currently telegraf triggers all input plugins on startup which generates a lot of load on 3rd party systems (and a lot of duplicated data)
### Current behavior:

Input plugins are triggered on startup and then every $interval.
### Desired behavior:

Add an option to the plugin configuration:

```
[[input.pluginname]]
interval="24h"
offset = "+3h"
```

the plugin will be started at 3am

```
[[input.pluginname]]
interval="1h"
offset = "-10m"
```

the plugin will be started at 0:50, 1:50, 2:50,...
### Use case:
- I have a plugin that collects 50k measurements. If this plugin is triggered at every startup of telegraf, the output TSDB gets flooded with lots of duplicate data (the data doesn't change more than once per 24h, but the TSDB doesn't know this as the timestamp is updated)

### LINKED ISSUE #10406: displacement on round_interval for improved remote monitoring
## Feature Request

Make it possible to start checking on intervals that are NOT round_interval - most useful for measuring remote response.

### Proposal:

Implement a displacement that will be added to round_interval. 

### Current behavior:

```
interval = "10s"
collection_jitter = "0s"
round_interval = true
```

causes collection to start at `:00`, `:10`, `:20` etc

### Desired behavior:

```
interval = "10s"
displacement = "5s"
collection_jitter = "0s"
round_interval = true
```

causes collection to start at `:05`, `:15`, `:25` etc

### Use case:

To provide the measurements for SLA reporting, we have several servers used for remote monitoring.

To simplify, assume a customers web-service. To monitor it we configure telegraf on two different servers.
Both of these telegraf polls the web-service. The reason for having two servers is that we always have monitoring of the service - even when after patching the telegraf-servers (obviously assuming that we do not patch them at the same time), or having any other downtime on one of the telegraf-servers.

In normal situations (i.e. when both telegraf-servers are up), we would like to poll the web-service (almost) equidistantly. This cannot be achieve with todays configuration. With a `displacement` that would be easy and this would also reduce the time to detect that a service is down.

### PR DESCRIPTION
**Pull request by srebhan**
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #1319
resolves #10406 

related to #10530

This PR addresses a long standing feature-request to be able to define an offset from the collection interval. This can be helpful in order to manually schedule data collection between agents, if the global option is used with multiple agents, or, if the plugin option is used, between plugins. The new `collection_offset` allows to define a time offset from the (rounded and unaligned) collection time. The option is valid in either the agent section or in any input-plugin section (as other interval parameters).

### CODE REVIEW

File: agent/tick.go [hunk 1]
```diff
@@ -75,6 +77,7 @@ func (t *AlignedTicker) next(now time.Time) time.Duration {
 		d = t.interval
 	}
 	d += internal.RandomDuration(t.jitter)
+	d += t.offset
```
Comment (Hipska) [inline agent/tick.go:80]: To me, it looks more logical to switch these. In my mind i'm doing offset+jitter instead of jitter+offset.

**Comment by sjwang90** [comment]
Tested the artifacts and everything works as expect. Awesome work @srebhan 😄

**Comment by poulhs** [comment]
I have tested a few different scenarios and it works **exactly** as expected. Very nice and very useful.

My only comment is about the comment... I think that "`manually scheduling`" may be a bit obfuscating - it does not require anyone to hit return to continue... ;-)

**Comment by reimda** [comment]
Hi Sven, it seems like 1319 and 10406 are requesting exact timing based on the system clock, kind of like cron can trigger on x minute of the hour. In those issues the use case is to have telegraf instances on multiple machines have synchronized intervals. (for example machine 1 gathers on the first minute of every hour and machine 2 gathers on the 30th minute of every hour)

10530 and this PR add an offset to telegraf's current interval. This doesn't seem to be the same thing as synchronizing the interval to the clock. Am I missing something? Maybe you intend people to use external tools like cron or at to start telegraf at an exact time?

**Comment by poulhs** [comment]
I think you missed something, @reimda - To get exact timing you must control more configurations.
Assume you want to telegraf to perform a check every 10 seconds, at :05, :15, :25, ... you do:

- `interval = "10s" ` # every 10 seconds
- `collection_jitter = "0s" ` # "unrandom" - doing exact timing:
- `round_interval = true ` # relative to systemclock - not starttime:
- `collection_offset = "5s"` ` # and changing from :00, :10, :20, etc, to 05, :15, :25, etc:

That will do the trick - having the servers with different `collection_offset`s makes it possible to control this (one server as above, another with `collection_offset = "0s"``)

On a single telegraf, this can be used to ensure that multiple checks are started at different times "don't step on my toes" - just apply same `interval` and different `collection_offset`s
 
And I guess there exist several other  use-cases.... (time will show ☺)

**Comment by reimda** [comment]
Thanks @poulhs for the clarification. I misunderstood how round_interval works. This looks very useful to me now!

**Review by reimda — APPROVED** [review]

**Review by Hipska — APPROVED** [review]

**Review by powersj — APPROVED** [review]

**Comment by kaunix** [comment]
so are these changes available in 1.26?

**Comment by powersj** [comment]
> so are these changes available in 1.26?
Yes

**Comment by kaunix** [comment]
so I tried to use them in the conf file (see below) and telegraf wouldn't start:

[[inputs.exec]]
commands=["/usr/bin/iperf3  -O 7 -c 192.168.1.1 --json"]
#interval = "1783s"
interval = "1800s"
collection_jitter= "0s"
round_interval = true
collection_offset = "54s"
timeout = "40s"
data_format = "json"
json_query = "end"
name_override = "server.example.com-write"

[[inputs.exec]]
commands=["/usr/bin/iperf3 -O 7 -c 192.168.1.1 -R --json"]
#interval = "1801s"
interval = "1800s"
collection_jitter= "0s"
round_interval = true
collection_offset = "1"
timeout = "40s"
data_format = "json"
json_query = "end"
name_override = "server.example.com-read"

[[inputs.exec]]
commands=["/usr/bin/iperf3 -O 7 -c 192.168.1.2 --json"]
#interval = "1831s"
interval = "1800s"
collection_jitter= "0s"
round_interval = true
collection_offset = "15"
timeout = "40s"
data_format = "json"
json_query = "end"
name_override = "server2.example.com-write"

[[inputs.exec]]
commands=["/usr/bin/iperf3 -O 7 -c 192.168.1.2 -R --json"]
#interval = "1867s"
interval = "1800s"
collection_jitter= "0s"
round_interval = true
collection_offset = "35"
timeout = "40s"
data_format = "json"
json_query = "end"
name_override = "server2.example.com-read"

**Comment by powersj** [comment]
File a bug please, asking on a closed PR isn't the place

**Pull request merged**